Do not warn even if faraday-retry gem is missing

akihikodaki opened the pull request:
It makes little sense to emit a warning saying "to use retry middleware" when the user does not request that. Remove the warning.

Resolves #1567

----

### Before the change?

* It emits a warning saying "to use retry middleware..."

### After the change?
<!-- Please describe the behavior or changes that are being added by this PR. -->

* The warning is gone.

### Pull request checklist
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been reviewed and added / updated if needed (for bug fixes / features)

### Does this introduce a breaking change?
<!-- If this introduces a breaking change make sure to note it here any what the impact might be -->

Please see our docs on [breaking changes](https://github.com/octokit/.github/blob/master/community/breaking_changes.md) to help!

- [ ] Yes
- [X] No

----

pbstriker38 commented on lib/octokit/default.rb:
```diff
@@ -7,11 +7,6 @@
-  begin
-    require 'faraday/retry'
-  rescue LoadError
-    Octokit::Warnable.octokit_warn 'To use retry middleware with Faraday v2.0+, install `faraday-retry` gem'
```
This warning is the only thing that would even let you know that this gem does retries. There is nothing in the README about this. This would be better as an explicit option to enable retries.

akihikodaki replied:
So do you suggest disabling retries by default?

pbstriker38 replied:
That would be the best way to remove the warning. The retry functionality needs to be better documented.

akihikodaki replied:
Added commit 3395ed41c0ee7dbc39856c96599acee181638b24 to remove the retry middleware from the default middleware stack and commit c892eb3ad3e5e32e79c537f94df102947eaa92a2 to provide the retry middleware to insert it into the middleware stack explicitly.

pbstriker38 commented on README.md:
````diff
@@ -631,6 +631,22 @@ x-ratelimit-reset: "1377205443"
+Next, insert `Octokit::Middleware::Retry` before `Octokit::Response::RaiseError`:
+
+```ruby
+Octokit.middleware.insert Octokit::Response::RaiseError, Octokit::Middleware::Retry
````
I would think configuring it on the client would be better
```ruby
Octokit::Client.new(auto_retry: true)
```

and then in the connection sawyer options you can insert the retry middleware if @auto_retry
```ruby
conn_opts[:builder].insert(0, *retry_middleware) if @auto_retry
```

akihikodaki replied:
While having a parameter for `Octokit::Client.new` is more concise, there are a few advantages of explicitly requiring to interact with middleware:
1. A user can specify additional options.
2. It is clear that it interacts with middleware.
3. It is consistent with the suggestion for caching, which follows this section in `README.md`.

So there is some trade-off. What do you prefer?

pbstriker38 replied:
What happens if you want one client that retries and another that doesn't. Setting on Octokit.middleware is global

akihikodaki replied:
You can also configure an individual instance; see: https://github.com/octokit/octokit.rb/pull/1706/files#diff-7ee18df1929eee9ac41b3ffa5f60cc4904aeb944ff7be6034721b1a76dc2798dR520

akihikodaki committed:
Do not warn even if faraday-retry gem is missing

It makes little sense to emit a warning saying "to use retry middleware"
when the user does not request that. Remove the warning.

akihikodaki committed:
Duplicate the default middleware stack

Duplicate the default middleware stack so that it won't be modified.

akihikodaki committed:
Add Octokit::Middleware::Retry

Octokit::Middleware::Retry allows incorporating the retry logic without
relying on the default middleware stack.

akihikodaki committed:
Avoid retrying by default

No retry middleware is included in Faraday 2.x and retrying requires
Faraday Retry gem. The default middlware stack used to include a retry
middleware only when available, which means the behavior of Octokit
implicitly changes depending on to the Faraday version or the presence
of Faraday Retry gem.

Remove the retry middleware from the default middleware stack to avoid
such an implicit behavioral change.

akihikodaki commented:
Rebased to v10.0.0.
